crates/bevy_ecs/src/storage/thin_array_ptr.rs

```diff
@@ -0,0 +1,250 @@
+use crate::query::DebugCheckedUnwrap;
+use std::alloc::{alloc, handle_alloc_error, realloc, Layout};
+use std::num::NonZeroUsize;
+use std::ptr::NonNull;
+
+/// Similar to [`Vec<T>`], but with the capacity and length cut out for performance reasons.
+///
+pub struct ThinArrayPtr<T> {
+    data: NonNull<T>,
```

Comment:
I added capacity in debug mode. but because the type doesn't support push / extend methods, tracking the length means changing api, so I decided against it.